algosdk/atomic_transaction_composer.py

```diff
@@ -0,0 +1,529 @@
+    BUILT = 1
+
+    # SIGNED indicates that the atomic group has been finalized and signed,
+    # but not yet submitted to the network.
+    SIGNED = 2
+
+    # SUBMITTED indicates that the atomic group has been finalized, signed,
+    # and submitted to the network.
+    SUBMITTED = 3
+
+    # COMMITTED indicates that the atomic group has been finalized, signed,
+    # submitted, and successfully committed to a block.
+    COMMITTED = 4
+
+
+class AtomicTransactionComposer:
+    """
+    Constructs an atomic transaction group which may contain a combination of
+    Transactions and ABI Method calls.
+
+    Args:
+        status (AtomicTransactionComposerStatus): IntEnum representing the current state of the composer
+        method_dict (dict): dictionary of an index in the transaction list to a Method object
+        txn_list (list[TransactionWithSigner]): list of transactions with signers
+        signed_txns (list[SignedTransaction]): list of signed transactions
+        tx_ids (list[str]): list of individual transaction IDs in this atomic group
+    """
+
+    MAX_GROUP_SIZE = 16
+
+    def __init__(self) -> None:
+        self.status = AtomicTransactionComposerStatus.BUILDING
+        self.method_dict = {}
+        self.txn_list = []
+        self.signed_txns = []
+        self.tx_ids = []
+
+    def get_status(self):
+        """
+        Returns the status of this composer's transaction group.
+        """
+        return self.status
+
+    def get_tx_count(self):
+        """
+        Returns the number of transactions currently in this atomic group.
+        """
+        return len(self.txn_list)
+
+    def clone(self):
+        """
+        Creates a new composer with the same underlying transactions.
+        The new composer's status will be BUILDING, so additional transactions
+        may be added to it.
+        """
+        cloned = AtomicTransactionComposer()
+        cloned.method_dict = copy.deepcopy(self.method_dict)
+        cloned.txn_list = copy.deepcopy(self.txn_list)
+        for t in cloned.txn_list:
+            t.txn.group = None
+        cloned.status = AtomicTransactionComposerStatus.BUILDING
+        return cloned
+
+    def add_transaction(self, txn_and_signer):
+        """
+        Adds a transaction to this atomic group.
+
+        An error will be thrown if the composer's status is not BUILDING,
+        or if adding this transaction causes the current group to exceed
+        MAX_GROUP_SIZE.
+
+        Args:
+            txn_and_signer (TransactionWithSigner)
+        """
+        if self.status != AtomicTransactionComposerStatus.BUILDING:
+            raise error.AtomicTransactionComposerError(
+                "AtomicTransactionComposer must be in BUILDING state for a transaction to be added"
+            )
+        if len(self.txn_list) == self.MAX_GROUP_SIZE:
+            raise error.AtomicTransactionComposerError(
+                "AtomicTransactionComposer cannot exceed MAX_GROUP_SIZE transactions"
```

Comment:
```suggestion
                f"AtomicTransactionComposer cannot exceed MAX_GROUP_SIZE ({MAX_GROUP_SIZE}) transactions"
```
could be helpful to add the actual max in an `f-string`

Reply:
Doesn't work now that I know we have to support 3.5. Will bring this up in standup.